seacatauth/credentials/providers/mysql.py

```diff
@@ -0,0 +1,386 @@
+import logging
+from typing import Optional
+
+import asab
+import aiomysql
+import passlib.hash
+import pymysql
+
+from .abc import EditableCredentialsProviderABC
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class MySQLCredentialsService(asab.Service):
+
+	def __init__(self, app, service_name="seacatauth.credentials.mysql"):
+		super().__init__(app, service_name)
+
+	def create_provider(self, provider_id, config_section_name):
+		return MySQLCredentialsProvider(self.App, provider_id, config_section_name)
+
+
+class MySQLCredentialsProvider(EditableCredentialsProviderABC):
+	# TODO: Use bind variables (https://legacy.python.org/dev/peps/pep-0249/#paramstyle)
+
+	Type = "mysql"
+
+	ConfigDefaults = {
+		"editable": "no",
+		"host": "localhost",
+		"port": "3306",
+		"database": "auth",
+		"table": "users",
+		"user": "root",
+		"password": "",
+		"field_id": "id",
+		"field_username": "username",
+		"field_email": "email",
+		"field_phone": "phone",
+		"field_password": "password",
+		"field_suspended": "suspended",
+		"data_fields": ""
+	}
+
+
+	def __init__(self, app, provider_id, config_section_name):
+		super().__init__(provider_id, config_section_name)
+		self.Editable = self.Config.getboolean("editable")
+		self.ConnectionParams = {
+			"host": self.Config.get("host"),
+			"port": self.Config.getint("port"),
+			"db": self.Config.get("database"),
+			"user": self.Config.get("user"),
+		}
+		password = self.Config.get("password")
+		if len(password) > 0:
+			self.ConnectionParams["password"] = password
+
+		self.Table = self.Config.get("table")
+
+		self.Fields = {
+			"username": self.Config.get("field_username"),
+			"email": self.Config.get("field_email"),
+			"phone": self.Config.get("field_phone"),
+			"suspended": self.Config.get("field_suspended"),
+		}
+		self.IdField = self.Config.get("field_id")
+		self.PasswordField = self.Config.get("field_password")
+
+		data_fields = self.Config.get("data_fields")
+		if len(data_fields) > 0:
+			self.DataFields = data_fields.split(" ")
+		else:
+			self.DataFields = None
+
+
+	async def create(self, credentials: dict) -> Optional[str]:
```

Comment:
Start with the read-only provider first.

Reply:
Provider is read-only by default. Set `editable=yes` to enable editing.